### ISSUE TITLE
Hive UDF java.lang.ClassNotFoundException: org.apache.logging.log4j.LogManager

### ISSUE THREAD

**Issue by iurnah** [nielsbasjes/yauaa#294]
**Describe the bug**
When add jar file yauaa-hive-6.0-udf.jar in hive and run the example query

got the following error.

`java.lang.NoClassDefFoundError: org/apache/logging/log4j/LogManager`

after add log4j2 as follows, same error.

hive (default)> ADD JAR log4j-core-2.14.1.jar
              > ADD JAR log4j-api-2.14.1.jar

the full log for reproduce the bug
```
hive (default)> ADD JAR yauaa-hive-6.0-udf.jar;
Added [yauaa-hive-6.0-udf.jar] to class path
Added resources: [yauaa-hive-6.0-udf.jar]
hive (default)> ADD JAR apache-log4j-2.14.1-bin/log4j-core-2.14.1.jar;
Added [apache-log4j-2.14.1-bin/log4j-core-2.14.1.jar] to class path
Added resources: [apache-log4j-2.14.1-bin/log4j-core-2.14.1.jar]
hive (default)> ADD JAR apache-log4j-2.14.1-bin/log4j-api-2.14.1.jar;
Added [apache-log4j-2.14.1-bin/log4j-api-2.14.1.jar] to class path
Added resources: [apache-log4j-2.14.1-bin/log4j-api-2.14.1.jar]
hive (default)> DROP  FUNCTION IF EXISTS parseuseragent;
OK
Time taken: 0.025 seconds
hive (default)> CREATE FUNCTION ParseUserAgent AS 'nl.basjes.parse.useragent.hive.ParseUserAgent';
OK
Time taken: 0.024 seconds
hive (default)> DESCRIBE FUNCTION EXTENDED ParseUserAgent;
OK
ParseUserAgent(str) - Parses the UserAgent into all possible pieces.
Synonyms: default.parseuseragent
Example:
> SELECT ParseUserAgent(useragent).DeviceClass,
         ParseUserAgent(useragent).OperatingsystemNameVersion,
         ParseUserAgent(useragent).AgentNameVersionMajor
  FROM   clickLogs;
+---------------+-----------------------------+------------------------+
|  deviceclass  | operatingsystemnameversion  | agentnameversionmajor  |
+---------------+-----------------------------+------------------------+
| Phone         | Android 6.0                 | Chrome 46              |
| Tablet        | Android 5.1                 | Chrome 40              |
| Desktop       | Linux Intel x86_64          | Chrome 59              |
| Game Console  | Windows 10.0                | Edge 13                |
+---------------+-----------------------------+------------------------+

Time taken: 0.02 seconds, Fetched: 16 row(s)
hive (default)> SELECT ParseUserAgent('Mozil la/5.0 (Linux\; Android 6.0\; Nexus 6 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.76 Mobile Safari/537.36');
Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/logging/log4j/LogManager
	at nl.basjes.parse.useragent.AbstractUserAgentAnalyzerDirect.<clinit>(AbstractUserAgentAnalyzerDirect.java:173)
	at nl.basjes.parse.useragent.hive.ParseUserAgent.initialize(ParseUserAgent.java:95)
	at org.apache.hadoop.hive.ql.udf.generic.GenericUDF.initializeAndFoldConstants(GenericUDF.java:139)
	at org.apache.hadoop.hive.ql.plan.ExprNodeGenericFuncDesc.newInstance(ExprNodeGenericFuncDesc.java:232)
	at org.apache.hadoop.hive.ql.parse.TypeCheckProcFactory$DefaultExprProcessor.getXpathOrFuncExprNodeDesc(TypeCheckProcFactory.java:959)
	at org.apache.hadoop.hive.ql.parse.TypeCheckProcFactory$DefaultExprProcessor.process(TypeCheckProcFactory.java:1176)
	at org.apache.hadoop.hive.ql.lib.DefaultRuleDispatcher.dispatch(DefaultRuleDispatcher.java:90)
	at org.apache.hadoop.hive.ql.lib.DefaultGraphWalker.dispatchAndReturn(DefaultGraphWalker.java:94)
	at org.apache.hadoop.hive.ql.lib.DefaultGraphWalker.dispatch(DefaultGraphWalker.java:78)
	at org.apache.hadoop.hive.ql.lib.DefaultGraphWalker.walk(DefaultGraphWalker.java:132)
	at org.apache.hadoop.hive.ql.lib.DefaultGraphWalker.startWalking(DefaultGraphWalker.java:109)
	at org.apache.hadoop.hive.ql.parse.TypeCheckProcFactory.genExprNode(TypeCheckProcFactory.java:193)
	at org.apache.hadoop.hive.ql.parse.TypeCheckProcFactory.genExprNode(TypeCheckProcFactory.java:146)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.genAllExprNodeDesc(SemanticAnalyzer.java:10941)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.genExprNodeDesc(SemanticAnalyzer.java:10896)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.genSelectPlan(SemanticAnalyzer.java:3762)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.genSelectPlan(SemanticAnalyzer.java:3544)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.genPostGroupByBodyPlan(SemanticAnalyzer.java:8813)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.genBodyPlan(SemanticAnalyzer.java:8768)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.genPlan(SemanticAnalyzer.java:9990)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.genPlan(SemanticAnalyzer.java:9883)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.genOPTree(SemanticAnalyzer.java:10522)
	at org.apache.hadoop.hive.ql.parse.CalcitePlanner.genOPTree(CalcitePlanner.java:293)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.analyzeInternal(SemanticAnalyzer.java:10533)
	at org.apache.hadoop.hive.ql.parse.CalcitePlanner.analyzeInternal(CalcitePlanner.java:199)
	at org.apache.hadoop.hive.ql.parse.BaseSemanticAnalyzer.analyze(BaseSemanticAnalyzer.java:223)
	at org.apache.hadoop.hive.ql.Driver.compile(Driver.java:446)
	at org.apache.hadoop.hive.ql.Driver.compile(Driver.java:312)
	at org.apache.hadoop.hive.ql.Driver.compileInternal(Driver.java:1201)
	at org.apache.hadoop.hive.ql.Driver.runInternal(Driver.java:1296)
	at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1127)
	at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1115)
	at org.apache.hadoop.hive.cli.CliDriver.processLocalCmd(CliDriver.java:224)
	at org.apache.hadoop.hive.cli.CliDriver.processCmd(CliDriver.java:176)
	at org.apache.hadoop.hive.cli.CliDriver.processLine(CliDriver.java:389)
	at org.apache.hadoop.hive.cli.CliDriver.executeDriver(CliDriver.java:754)
	at org.apache.hadoop.hive.cli.CliDriver.run(CliDriver.java:699)
	at org.apache.hadoop.hive.cli.CliDriver.main(CliDriver.java:634)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.hadoop.util.RunJar.run(RunJar.java:244)
	at org.apache.hadoop.util.RunJar.main(RunJar.java:158)
Caused by: java.lang.ClassNotFoundException: org.apache.logging.log4j.LogManager
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 44 more
```

**Component where the bug happens**
[ ] Core analyzer
[X] UDF : hive (pig have similar NoClassDefFoundError)
[ ] Other

**To Reproduce**
Steps or code fragment to reproduce the behavior:

See the full logs above.

**Expected behavior**
successfully run the hive query

**Screenshots**
N/A

**Additional context**
N/A

**Comment by nielsbasjes** [comment]
@iurnah Just quick check; Which Java (JRE) and Hive versions are you on?

**Comment by paulmackinlay** [comment]
The problem is https://github.com/nielsbasjes/yauaa/blob/master/analyzer/src/main/java/nl/basjes/parse/useragent/AbstractUserAgentAnalyzerDirect.java#L51
There is a specific import for log4j, I upgraded to version 6.0 and this is causing me an issue as well. Probably you should import java.util.logging.LogManager

**Comment by iurnah** [comment]
Hi @nielsbasjes, please find more info about my environment below.

```bash
$ java -version
java version "1.8.0_172"
Java(TM) SE Runtime Environment (build 1.8.0_172-b11)
$ hive --version
Hive 1.1.0.239
```

**Comment by iurnah** [comment]
> The problem is https://github.com/nielsbasjes/yauaa/blob/master/analyzer/src/main/java/nl/basjes/parse/useragent/AbstractUserAgentAnalyzerDirect.java#L51
> There is a specific import for log4j, I upgraded to version 6.0 and this is causing me an issue as well. Probably you should import java.util.logging.LogManager

Thanks, @paulmackinlay. I'll give it a try. Do you know why `import org.apache.logging.log4j.LogManager;` doesn't work? If you have a fix, do you want to contribute by send a pull request? Thank you very much!

**Comment by paulmackinlay** [comment]
@iurnah - in my case I don't use log4j, so when this class loads I end up getting a ClassNotFoundException because I don't have log4j in the classpath. Forcing users to depend on a particular logging implementation isn't great, really users should be able to choose whichever implementation they want.

SLF4J is implementation agnostic and will detect your chosen implementation at runtime, so it it useful for decoupling your code from logging frameworks (but it does force you to depend on SLF4J). I think using the classes that ship with the JDK is also an ok solution, I haven't checked but java.util.logging.LogManager may also detect a logging implementation at runtime. 

However for your code, if you are not bothered about depending on log4j then adding it to your classpath will be the quickest fix. It means you may need to manage different log files/logging configs though.

**Comment by paulmackinlay** [comment]
I checked out the code and have gone through it. SLF4J is already being used so the fix is to replace across the whole project these imports:

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

with:
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

and then replace all these calls:
LogManager.getLogger(...)
with
LoggerFactory.getLogger(...)

**Comment by nielsbasjes** [comment]
With the 6.0 version I "Fully replaced SLF4J with Log4J2" (hence the change from version 5.x to 6.x).

The reason you still see slf4j is because this library has dependencies (like org.apache.httpcomponents.client5:httpclient5 ) that require slf4j. I would really prefer to have only a single logging framework; but that is unlikely to happen.

So I spent some time getting a "full" Hadoop/Hdfs/Hive cluster running using docker-compose that simply tests the basics.
I based this code on some existing scripting I found that runs a full hive in docker.

https://github.com/nielsbasjes/yauaa/tree/master/udfs/hive/src/it

This uses Hive 2.3.2 on Hadoop 2.7.4 on java 8 and when I run this (with the published 6.0 version and the current code base) it all works as expected.

I did a cross check with the original report and I found these notable points:
1) This was reported on a very very old version of Hive: 1.1.0. I checked the hive website and version 1.1.0 was released on 8 March 2015. I'm not going to put effort in checking against Hive 1.x.
2) I noticed that the original report shows `ADD JAR yauaa-hive-6.0-udf.jar;`, this to me seems like the jar file that was added was local to the client which seems to me like a likely source of problems when you run it in a cluster. Also this is a relative path! (Relative to where the client was running).

In my testing code (and also in my documentation https://yauaa.basjes.nl/UDF-ApacheHive.html) I upload the jar file to HDFS and then specify the absolute path like this ```
ADD JAR hdfs:///udf/yauaa-hive-6.0-udf.jar;
CREATE   FUNCTION ParseUserAgent AS 'nl.basjes.parse.useragent.hive.ParseUserAgent';
```
or 
```
CREATE   FUNCTION ParseUserAgent AS 'nl.basjes.parse.useragent.hive.ParseUserAgent' USING JAR 'hdfs:///udf/yauaa-hive-6.0-udf.jar';
```

Doing this makes it work on my machine perfectly.

Please double check if this helps.

**Comment by nielsbasjes** [comment]
Also this works on a clean Hive (using this docker based setup).

```
CREATE FUNCTION ParseUserAgent
AS 'nl.basjes.parse.useragent.hive.ParseUserAgent'
USING JAR 'ivy://nl.basjes.parse.useragent:yauaa-hive:6.0?classifier=udf';
```

I also checked 'forcing' it to run an actual MR job on the Hadoop cluster by adding a `GROUP BY` clause.
```
SELECT ParseUserAgent(useragent).DeviceClass,
       ParseUserAgent(useragent).OperatingsystemNameVersion,
       ParseUserAgent(useragent).AgentNameVersionMajor
FROM   clickLogs 
GROUP BY useragent;
```

**Comment by nielsbasjes** [comment]
I noticed that Apache Pig was also mentioned.
I (manually) added Apache Pig to this docker image and tried to run this example https://github.com/nielsbasjes/yauaa/tree/master/examples/pig . 
Indeed the problem occurred `Caused by: java.lang.ClassNotFoundException: org.apache.logging.log4j.LogManager`.

This was easily fixed by doing `cp /opt/hive/lib/log4j-* /opt/pig-0.17.0/lib` which put these files in the lib directory of Pig.

    log4j-1.2-api-2.6.2.jar
    log4j-api-2.6.2.jar
    log4j-core-2.6.2.jar  
    log4j-jul-2.5.jar 
    log4j-slf4j-impl-2.6.2.jar 
    log4j-web-2.6.2.jar

With this simple addition of the dependencies is works on my machine.

**Comment by paulmackinlay** [comment]
@nielsbasjes  thanks for the clarification. It's unusual to replace SLF4J which gives the flexibility to use any logging implementation with one specific logging implementation, however I can see in your pom that you still depend on SLF4J and your dependency on log4j is 'provided' which means your project will not easily be used for projects that don't use log4j, it ends up causing runtime exceptions. If you do want to move completely to log4j you are better off removing all references to SLF4J and make log4j a compile dependency.

Just out of interest why move away from SLF4J for a particular implementation? That makes adopting your code quite a bit more difficult. In my case I simply won't be able to use it, I'll have to change your code to be SLF4J based.

Is your intention to use log4j as your **preferred** logging implementation rather than forcing all adopters of your code to use log4j? If that is what you want to do I'm happy to help.

**Comment by nielsbasjes** [comment]
The nasty thing about logging frameworks is that there are too many.
This is a library intended to be used in all kinds of scenarios and I do not want to impose any logging framework.
I would prefer having a single "facade" like Slf4J intends to be.

The problem is that I use several dependencies that use a different logging framework/facade for themselves.

The basic idea of SLF4J is to have a single frontend that plugs in into any backend you prefer.

The big problem I saw is that SLF4J had not seen any updates / commits for a long time. Between december 2019 and june 2021 there were no significant commits at all (only minor typos). I came to the conclusion (at that point in time) this project had unfortunately died.

Now log4J does the exact same thing (it is also a logging facade), was already a dependency in my library and is updated on a contineous basis.
https://logging.apache.org/log4j/2.x/ 
 `The Log4j API is a logging facade that may, of course, be used with the Log4j implementation, but may also be used in front of other logging implementations such as Logback. `

So I "simply" switched to log4j2 and apparently the 2.x is not a dependency that is readily available in older platforms like the hive 1.x on which is issue was reported.

The solution for those older platforms is to make sure the required libraries are available on the classpath.

If I can I would like to hide (shade perhaps) all of the JCL and slf4j code and ONLY have the log4j2 as a logging dependency.
I created a separate issue for that: https://github.com/nielsbasjes/yauaa/issues/303
